### ISSUE TITLE
Development plan for ESPnet2 speech enhancement 

### ISSUE THREAD

**Issue by LiChenda** [espnet/espnet#2200]
We have been developing (enhancing) speech enhancement functions in ESPnet2.
We particularly design it to tightly integrate it with downstream speech processing tasks (I.e., ASR and ST). 
Phase 1 is to implement a stand alone basic speech enhancement and separation techniques as a reference technique. In Phase 2, they are flexibly called as a front-end processing subnetwork from ASR/ST/SLU/DIAR with maintaining computational graphs for joint training.

**We also welcome your requests for this particular directions and contributions!**

## Progress & Plan  #2124 

- [x] A new task for speech enhancement
- [x] Speech enhancement recipe template
- [x] Evaluate with pretrained ASR
- [x] Joint training with ASR/ST/SLU/DIAR tasks
- [x] Single-channel multi-talker ASR
- [ ] Support integration of models from other tools 
    - [ ] [PaderTorch](https://github.com/fgnt/padertorch)
    - [x] [Asteroid](https://github.com/mpariente/asteroid)

#### frequency domain
- [x] Conformer
- [x] [DCCRN](https://www.isca-speech.org/archive/interspeech_2020/hu20g_interspeech.html)
- [x] [Deep Attractor Network](https://arxiv.org/abs/1611.08930)
- [x] [Deep Clustering](https://ieeexplore.ieee.org/abstract/document/7471631/) series
- [ ] [DPT-FSNet](https://arxiv.org/abs/2104.13002)
- [ ] [Speaker-Conditional Chain Model](https://www.isca-speech.org/archive/interspeech_2020/shi20e_interspeech.html)
- [ ] [TFPSNet](https://ieeexplore.ieee.org/document/9747554/)
- [x] [TF-GridNet](https://arxiv.org/abs/2209.03952)
- [x] Transformer
- [x] [uPIT](https://ieeexplore.ieee.org/abstract/document/7979557/)

#### time domain
- [ ] [CDiffuSE](https://arxiv.org/abs/2202.05256)
- [x] [Conv-TasNet ](https://arxiv.org/abs/1809.07454)
- [x] [DPRNN](http://arxiv.org/abs/1910.06379v2)
- [x] [DPTNet](https://arxiv.org/abs/2007.13975)
- [ ] [MixIT](https://arxiv.org/abs/2006.12701) (PRing, see #4263)
- [x] [SkiM](https://arxiv.org/abs/2201.10800)
- [x] [SVoice](https://arxiv.org/abs/2003.01531)
- [ ] [Wave-U-Net](https://arxiv.org/abs/1806.03185)
- [ ] [Wavesplit](http://arxiv.org/abs/2002.08933v2)

#### frequency domain
- [ ] DNN-based AuxIVA-ISS (PRing, see #4201)
- [x] [DC_CRN](https://web.cse.ohio-state.edu/~wang.77/papers/TZW.taslp21.pdf)
- [x] DNN-Beamformer
  - [x] MVDR
  - [x] MPDR
  - [x] wMPDR
  - [x] LCMV
  - [x] LCMP
  - [x] wLCMP
  - [x] GEV
  - [x] Multichannel Wiener Filter 
  - [x] [WPD](https://ieeexplore.ieee.org/abstract/document/8691481/)
  - [x] [time-frequency-bin-wise switching (TFS) MVDR](https://ieeexplore.ieee.org/document/8683528)
- [ ] Online beamformers
- [x] [DNN-WPE](https://ieeexplore.ieee.org/document/6255769)
- [x] [iNeuBe](https://arxiv.org/abs/2202.12298)
- [ ] [VarArray](https://arxiv.org/abs/2110.05745)

#### time domain
- [x] [FaSNet](https://arxiv.org/abs/1909.13387)
- [x] [iFaSNet](https://www.isca-speech.org/archive/interspeech_2021/luo21c_interspeech.html)

## Documentation 
- [ ] add documentation in https://github.com/espnet/espnet/tree/master/doc
- [ ] describe some features and functions in the main README.md
- [x] prepare the webpage to put the audio demonstration
- [x] and/or create notebook ([Google Colab](https://colab.research.google.com/drive/1fjRJCh96SoYLZPRxsjF9VDv4Q2VoIckI
))
## Recipes:
- [x] aishell4
- [x] clarity21
- [ ] chime1
- [ ] chime2
- [x] chime4
- [x] conferencingspeech21
- [ ] dirha_wsj
- [x] dns_ins20
- [x] dns_icassp21
- [x] dns_ins21
- [x] l3das22
- [ ] libricss
- [x] librimix
- [x] lt_slurp_spatialized
- [ ] reverb
- [x] sms_wsj
- [x] vctk_noisy
- [x] vctk_noisyreverb
- [x] wham
- [x] whamr
- [x] wsj0_2mix 
- [x] wsj0_2mix_spatialized

## Others

- [ ] more evaluation metrics: 
   - [x] SDR/SAR/SIR
   - [x] STOI
   - [x] extended STOI
   - [ ] PESQ (make it optional to user)
   - [x] ASR
- [ ] More loss functions
  - [x] SI-SNR
  - [x] CI-SDR
  - [x] mask loss
      - [x] IRM 
      - [x] IBM 
      - [x] IAM 
      - [x] PSM
      - [x] NPSM
      - [x] CRM
  - [x] spectrum loss
  - [x] [absolute coherence loss](https://arxiv.org/abs/2008.11273)

- [x] integration test
- [x] unit test
- [x] flexibly changing of the base network architecture, e.g., BLSTM <-> CNN <-> SA <-> Confromer   
- [x] unifying the encoder/decoder and stft/istft in tasnet and masking models.
- [x] add noise and reverberation on-the-fly
- [x] dynamic mixing for the speech separation task

**Comment by kamo-naoyuki** [comment]
Could you create a recipe of mini_an4/enh1 and add it in ci/test_integration.sh? I think you can create a recipe with num_spk=1.

**Comment by sw005320** [comment]
Please reflect https://github.com/espnet/espnet/pull/2212 to `enh.sh`

**Comment by kamo-naoyuki** [comment]
I also reflected it to enh.sh. We might update the system more, for now, you don't need to care it. I'll follow it.

**Comment by sw005320** [comment]
@kamo-naoyuki, thanks for taking care of it!

**Comment by Emrys365** [comment]
> Could you create a recipe of mini_an4/enh1 and add it in ci/test_integration.sh? I think you can create a recipe with num_spk=1.

It's now added in our develop branch.

**Comment by kamo-naoyuki** [comment]
@Emrys365 Could you merge it with the highest priority? I feel uneasy now when touching enhancement part.

**Comment by Emrys365** [comment]
Sure, I will make a PR soon.

**Comment by kamo-naoyuki** [comment]
I forgot to say. We are following google style docstring: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

This is not important, so if you'll modify any docstrings, please change to the style at that time.
(e.g. https://github.com/espnet/espnet/blob/b2d1d44668d44885d8a0ac496dd59136a8d1d1cc/espnet2/enh/espnet_model.py#L328-L333)

**Comment by b-flo** [comment]
@kamo-naoyuki There is another docstring style I'm using which was used at one point (dunno why we discarded it?):

```
"""time-frequency L1 loss. 
  
Arguments:
         ref (torch.Tensor):  (B, T, F) or (B, T, C, F) 
         inf (torch.Tensor) : (B, T, F) or (B, T, C, F) 

Returns:
         (torch.Tensor): (B) 

"""
```

Should I modify all files to match google style?

**Comment by kamo-naoyuki** [comment]
If you notice it, but you don't need to change using your time. 
(Actually we can replace all of them with something like sed command if we feel like it, so)

Also note that type information in docstring is not necessary, but type-annotation is more preferable.

**Comment by kamo-naoyuki** [comment]
I think it's more readable that all entity of scripts and pyscripts are put at `egs2/TEMPLATE/asr1`. I mean `egs2/TEMPLATE/enh1/scripts` is a little bit complex because there are many symbolic links to `egs2/TEMPLATE/asr1/`.

Yes, it's also confusing if `egs2/TEMPLATE/asr1/scripts` has the scripts for the other task, but even though, I think it's better to put them `egs2/TEMPLATE/asr1`.

What do you think?

**Comment by mpariente** [comment]
Just discovering this thread, that's very cool!

It's been a while that making an end-to-end Asteroid/ESPnet recipe is on our todo-list, would you be interested?

**Comment by sw005320** [comment]
Sure, we're definitely interested in it.
Thanks for reaching us!
Can you contact me by email (shinjiw@ieee.org)?

**Comment by deepconsc** [comment]
I'll be happy to integrate v2 version of our paper: https://arxiv.org/abs/2006.07637

**Comment by sw005320** [comment]
> I'll be happy to integrate v2 version of our paper: https://arxiv.org/abs/2006.07637

Sounds good.
We're now further refactoring our code and adding several examples.
So, you can wait for our development to be fixed in the near future.

**Comment by neillu23** [comment]
Hi, I am from Academia Sinica, Taiwan
I have discussed with Prof. Watanabe about the ESPnet-SE, finding the deep-feature-loss in our paper: https://arxiv.org/pdf/2011.07442.pdf possibly could add to ESPnet-SE.
Would you think this loss would fit in your system?

**Comment by sw005320** [comment]
Thanks!
We (me, @LiChenda, @Emrys365, and @shincling) will discuss how we can integrate your efforts.

**Comment by Alex-Songs** [comment]
Hello, where can I find the recipe corresponding to enh_asr.py? Thank you! @LiChenda

**Comment by shincling** [comment]
> Hello, where can I find the recipe corresponding to enh_asr.py? Thank you! @LiChenda

@Alex-Songs Alex, we are currently in the process of several individual PRs for Espnet-SE, recipes will be the main content of our next PR, please wait for our update in the near future. Thanks.

**Comment by Alex-Songs** [comment]
Thanks! @shincling

**Comment by Alex-Songs** [comment]
Hello, is there any related paper introduction for "STFT consistency"? Thank you!  @shincling

**Comment by Emrys365** [comment]
> Hello, is there any related paper introduction for "STFT consistency"? Thank you! @shincling

Hi, you can refer to the [paper](Explicit consistency constraints for STFT spectrograms and their application to phase reconstruction) by Jonathan Le Roux, and [a more recent paper](https://arxiv.org/pdf/1811.08521.pdf) by Scott Wisdom.
Our implementation is based on the latter one.

**Comment by Alex-Songs** [comment]
Thanks! @Emrys365

**Comment by Alex-Songs** [comment]
Hello, when I use tf-mask_net and ASR for joint training (2 speaker), the CTC loss function is difficult to converge. Have you encountered a similar situation? Thanks!

**Comment by shincling** [comment]
> Hello, when I use tf-mask_net and ASR for joint training (2 speaker), the CTC loss function is difficult to converge. Have you encountered a similar situation? Thanks!

Hi Alex. Could you tell me the script、recipe and configure file you used? Before, I have tried the tf-masking with the asr, but not works very well.  I could revisit it now.

**Comment by Alex-Songs** [comment]
Hello shincling, I used the default configuration of wsj0_2mix, but the training set was replaced with other simulated data sets. I found that TasNet works well, but tf-mask-net is difficult to converge. Thank you! Happy New Year in advance! @shincling

**Comment by Dahee96** [comment]
@Alex-Songs 
Hi Alex. Did you joint training TasNet and ASR?
Could you upload or share your code? I have difficulty in doing the process with other data :(

**Comment by kirianguiller** [comment]
Hi everyone, thanks for this great resources. I wanted to know where could I find the different recipes for speech enhancement (for single source). I can't find them on the doc of espnet2.

Thanks !

**Comment by LiChenda** [comment]
> Hi everyone, thanks for this great resources. I wanted to know where could I find the different recipes for speech enhancement (for single source). I can't find them on the doc of espnet2.
> 
> Thanks !

Hi， @kirianguiller ! It was in our develop branch and we recently made a recipe pull request (see #2888) to the main branch. I believe it will be merged in near future. You can find the single source enhancement in `CHiME4` recipe. `WHAM`, `WHAMR`, `LibriMix` also support single source data in the data preparing stage (By default, they are 2-speaker separartion recipes).

**Comment by kirianguiller** [comment]
Thanks @LiChenda for your reply. 

I can see in #2888 that you say : 
> > Pre-trained models?
> > There are some discussions in #2878
> 
> Since we have refactored the enhancement models in to new structure, the previous trained models are not compatible with the new codes. We are training new models for these recipes and thus the pretrained models are not included here. Once they are done, I'll upload the models. I think they shoud be done in this PR ? @sw005320

Should I deduce from this quote that no pretrained models are available for speech enhancement at the moment ? (I'm looking  for a SE denoising model :) )

**Comment by LiChenda** [comment]
> Should I deduce from this quote that no pretrained models are available for speech enhancement at the moment ? (I'm looking for a SE denoising model :) )

At this moment, the pretrained model for new code is not avaliable. If you need it now, here are some possible solutions:

- I can send you pretrained models (old version, CHiME4 or WHAM ) to you, and you can checkout the previous code from the history.
- Use the new version python code, convert and load the old pretrained model manually.
- If you have data and GPU, you can train a model yourself with the recipes. The default configuration worked in the pervious version, since the refactoring does not change the logic of the models, I believe they are OK.

**Comment by kirianguiller** [comment]
Thanks, I can try with an old version of a pretrained model. Which versions in the history are compatible ?

Btw, I have a GPU but it is being used for training something else at the moment. Maybe I can try to train a Chime4 enhancement model later in the week.

**Comment by LiChenda** [comment]
Hi, @kirianguiller , I have put the old models [here](https://drive.google.com/drive/folders/1Z9_rX-jej4MIDFoXLnUSoPVucfER9EoF?usp=sharing). Thanks to @Emrys365 for providing the models.
You can checkout this [branch](https://github.com/LiChenda/espnet/tree/develop_rebased) to use these models. If you have any other questions, feel free to email me lichenda1996@sjtu.edu.cn

**Comment by kirianguiller** [comment]
Thank you Li ! I will try this and will let you know how it is going :)

**Comment by navta** [comment]
Hi, I'm looking for a recipe to reproduce Table 6 in your "ESPNET-SE: END-TO-END SPEECH ENHANCEMENT AND SEPARATION TOOLKIT DESIGNED FOR ASR INTEGRATION" paper.

I know the joint training with ASR task is PRing here. Just wondering if there is a working recipe in another place.

Thanks !

**Comment by Emrys365** [comment]
> Hi, I'm looking for a recipe to reproduce Table 6 in your "ESPNET-SE: END-TO-END SPEECH ENHANCEMENT AND SEPARATION TOOLKIT DESIGNED FOR ASR INTEGRATION" paper.
> 
> I know the joint training with ASR task is PRing here. Just wondering if there is a working recipe in another place.
> 
> Thanks !

Hi @navta, you could find an old version of recipes [here](https://github.com/LiChenda/espnet/tree/develop_rebased).

**Comment by tz579** [comment]
Dear Authors,

Is there any working recipe(s) in the https://github.com/LiChenda/espnet/tree/develop_rebased that uses the neural beamformer as the SE model? Either joint SE-ASR or standalone SE recipe is OK for me.

Thanks!

**Comment by Emrys365** [comment]
> Dear Authors,
> 
> Is there any working recipe(s) in the https://github.com/LiChenda/espnet/tree/develop_rebased that uses the neural beamformer as the SE model? Either joint SE-ASR or standalone SE recipe is OK for me.
> 
> Thanks!

Hi @tz579 , sorry for the late response. You could run egs2/chime4/enh1 and egs2/whamr/enh1 (set training, validation, adn evaluation sets to "*_single_*") for beamforming based SE. The joint SE-ASR for single-speaker is not yet tested.

**Comment by tz579** [comment]
> > Dear Authors,
> > Is there any working recipe(s) in the https://github.com/LiChenda/espnet/tree/develop_rebased that uses the neural beamformer as the SE model? Either joint SE-ASR or standalone SE recipe is OK for me.
> > Thanks!
> 
> Hi @tz579 , sorry for the late response. You could run egs2/chime4/enh1 and egs2/whamr/enh1 (set training, validation, adn evaluation sets to "__single__") for beamforming based SE. The joint SE-ASR for single-speaker is not yet tested.

Dear Author,

Thanks for that! There is only one question now: I saw that the joint SE-ASR work has been merged into the default branch of https://github.com/LiChenda/espnet/ more than 3 months ago. When is the estimated time for it to be eventually merged here into the official repo?

Thanks a lot!

**Comment by Emrys365** [comment]
> Dear Author,
> 
> Thanks for that! There is only one question now: I saw that the joint SE-ASR work has been merged into the default branch of https://github.com/LiChenda/espnet/ more than 3 months ago. When is the estimated time for it to be eventually merged here into the official repo?
> 
> Thanks a lot!

Hi, thank you for your interests in the progress. We are intensively working on the joint training framework, and hopefully it can be completed by the end of this month.

**Comment by tz579** [comment]
> > Dear Author,
> > Thanks for that! There is only one question now: I saw that the joint SE-ASR work has been merged into the default branch of https://github.com/LiChenda/espnet/ more than 3 months ago. When is the estimated time for it to be eventually merged here into the official repo?
> > Thanks a lot!
> 
> Hi, thank you for your interests in the progress. We are intensively working on the joint training framework, and hopefully it can be completed by the end of this month.

Thanks a lot for letting me know!

Best,
